Core/src/org/sleuthkit/autopsy/mainui/datamodel/DataArtifactDAO.java

```diff
@@ -154,6 +174,94 @@ public void dropDataArtifactCache() {
+        ModuleDataEvent dataEvt = DAOEventUtils.getModuelDataFromArtifactEvent(evt);
+        if (dataEvt == null) {
+            return Collections.emptySet();
+        }
+
+        Map<BlackboardArtifact.Type, Set<Long>> artifactTypeDataSourceMap = dataEvt.getArtifacts().stream()
+                .map((art) -> {
+                    try {
+                        if (BlackboardArtifact.Category.DATA_ARTIFACT.equals(art.getType().getCategory())) {
+                            return Pair.of(art.getType(), art.getDataSourceObjectID());
+                        }
+                    } catch (TskCoreException ex) {
+                        logger.log(Level.WARNING, "Unable to fetch artifact category for artifact with id: " + art.getId(), ex);
+                    }
+                    return null;
+                })
+                .filter(pr -> pr != null)
+                .collect(Collectors.groupingBy(pr -> pr.getKey(), Collectors.mapping(pr -> pr.getValue(), Collectors.toSet())));
+
+        // don't do anything else if no relevant events
+        if (artifactTypeDataSourceMap.isEmpty()) {
+            return Collections.emptySet();
+        }
+
+        // invalidate cache entries that are affected by events
+        ConcurrentMap<SearchParams<BlackboardArtifactSearchParam>, DataArtifactTableSearchResultsDTO> concurrentMap = this.dataArtifactCache.asMap();
+        concurrentMap.forEach((k, v) -> {
+            Set<Long> dsIds = artifactTypeDataSourceMap.get(k.getParamData().getArtifactType());
+            if (dsIds != null) {
+                Long searchDsId = k.getParamData().getDataSourceId();
+                if (searchDsId == null || dsIds.contains(searchDsId)) {
+                    concurrentMap.remove(k);
+                }
+            }
+        });
+
+        // gather dao events based on artifacts
+        List<DataArtifactEvent> dataArtifactEvents = new ArrayList<>();
+        for (Entry<BlackboardArtifact.Type, Set<Long>> entry : artifactTypeDataSourceMap.entrySet()) {
+            BlackboardArtifact.Type artType = entry.getKey();
+            for (Long dsObjId : entry.getValue()) {
+                DataArtifactEvent newEvt = new DataArtifactEvent(artType, dsObjId);
+                dataArtifactEvents.add(newEvt);
+            }
+        }
+
+        List<TreeEvent> newTreeEvents = this.treeCounts.enqueueAll(dataArtifactEvents).stream()
+                .map(daoEvt -> new TreeEvent(getTreeItem(daoEvt.getArtifactType(), daoEvt.getDataSourceId(), TreeDisplayCount.INDETERMINATE), false))
+                .collect(Collectors.toList());
+
+        return Stream.of(dataArtifactEvents, newTreeEvents)
+                .flatMap((lst) -> lst.stream())
+                .collect(Collectors.toSet());
+    }
+
+    private TreeItemDTO<DataArtifactSearchParam> getTreeItem(BlackboardArtifact.Type artifactType, Long dataSourceId, TreeDisplayCount displayCount) {
```

Comment:
We should rename this to something like createTreeItem() since it is making something new versus retrieving.